presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/dereference/ExtractDereferencesFromFilterAboveScan.java

```diff
@@ -0,0 +1,89 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.prestosql.sql.planner.iterative.rule.dereference;
+
+import com.google.common.collect.BiMap;
+import com.google.common.collect.HashBiMap;
+import io.prestosql.matching.Capture;
+import io.prestosql.matching.Captures;
+import io.prestosql.matching.Pattern;
+import io.prestosql.sql.planner.ExpressionExtractor;
+import io.prestosql.sql.planner.Symbol;
+import io.prestosql.sql.planner.TypeAnalyzer;
+import io.prestosql.sql.planner.iterative.Rule;
+import io.prestosql.sql.planner.plan.Assignments;
+import io.prestosql.sql.planner.plan.FilterNode;
+import io.prestosql.sql.planner.plan.PlanNode;
+import io.prestosql.sql.planner.plan.ProjectNode;
+import io.prestosql.sql.planner.plan.TableScanNode;
+import io.prestosql.sql.tree.DereferenceExpression;
+import io.prestosql.sql.tree.ExpressionTreeRewriter;
+
+import static io.prestosql.matching.Capture.newCapture;
+import static io.prestosql.sql.planner.iterative.rule.dereference.PushDownDereferencesUtil.validDereferences;
+import static io.prestosql.sql.planner.plan.Patterns.filter;
+import static io.prestosql.sql.planner.plan.Patterns.source;
+import static io.prestosql.sql.planner.plan.Patterns.tableScan;
+import static java.util.Objects.requireNonNull;
+
+/**
+ * Extracts all dereference expressions from a filter node located above a table scan. This enables pushdown of
+ * dereference projections from the filter node into table scan using the {@link io.prestosql.sql.planner.iterative.rule.PushProjectionIntoTableScan} rule.
+ */
+public class ExtractDereferencesFromFilterAboveScan
```

Comment:
* The general approach followed in these rules is that we always push down the optimal dereferences. i.e. if we have a.x and a.y.z projected out of a node, we only push a.y down. This is to avoid duplicate shuffling/copying in the execution.   

* Following that logic, we can not extract such differences from  filternode itself, without a projection on top of it. This is because since output symbols of a filternode are the same as input ones. For example, consider the following:

FilterNode (a.x = 5)
	Node (a)

Then, just looking at this node, we won't be able to pushdown a.x since a might also be required up the tree, and we're duplicating the propagation of `a.x`. If there is an explicit Project on top of this FilterNode projecting a.x, we can push it down.

An example of why extracting dereferences from a FilterNode can be suboptimal:

SELECT T1.a FROM T1, T2 ON T1.x = T2.X WHERE T1.a.b > T2.y

Before

```
FilterNode(T1.a.b > T2.y)
	Join(T1.x = T2.x)
		T1 (x, a)
		T2 (y)
```

If we extract dereferences without projection, we end up with the following:

```
FilterNode(q > y)
	Join(T1.x = T2.x)
		Project (q := a.b, x, a)
			T1(x, a)

```

Here, we're replicating a.b in potential shuffle, which is not optimal. The more optimal thing  would be to not push down a.b. That's why we don't extract dereferences from filter.

If there is a Project(d:=a.b) on top of the FilterNode, it'd make sense to push it down. This is taken care of by PushDownDereferencesThroughFilter rule.

The case of ExtractDereferencesFromFilterAboveScan is special because we want to leverage applyProjection and applyFilter. Even if there's no projection above the FilterNode-->TableScan combo, we'd still like to pushdown the dereferences so that `applyProjection` can create virtual columns. And the next invocation of applyFilter would push the predicates down. 


I think I can summarize as the following comment in the code, if the explanation looks good to you:

* Extracting dereferences from a Filter (eg. FilterNode(a.x = 5)) can be suboptimal if columns are being accessed up the plan tree (eg. a), because it results in replication of fields (eg. a.x). So it's safer to pushdown dereferences from Filter only when there's an explicit projection on top of the filter node (Ref PushProjectionIntoTableScan).

* In case of a FilterNode on top of TableScanNode, we want to push all dereferences into a new ProjectNode below, so that PushProjectionIntoTableScan optimizer can push those columns in the connector, and provide new column handles for subcolumns. PushPredicateIntoTableScan optimizer can then push predicates on these subcolumns into the connector.

Reply:
Excellent explanation!